[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69060
* __->__ #69027

Resubmission of https://github.com/pytorch/pytorch/pull/62964 with the
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, the tests that we need to add are:

- [x] Gradient hooks are called once
- [x] works when input does require grads but Tensor that require grads are captures (like first layer in a nn)
- [x] works for functions with arbitrary input/output objects
- [x] distributed tests (next PR)

Note that this is only for `torch.utils.checkpoint`, if this approach overall looks good, we will do something similar for `checkpoint_sequential`. 


Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

rohan-varma committed:
[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd

Resubmission of https://github.com/pytorch/pytorch/pull/62964 withe
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, we have added
the following tests:

-[ ] Gradient hooks are called once

Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

[ghstack-poisoned]

rohan-varma commented on torch/utils/checkpoint.py:
```diff
@@ -275,3 +287,85 @@ def forward(input):
     return run_function(end + 1, len(functions) - 1, functions)(input)
+
+
+class Checkpoint(torch.nn.Module):
```
Implementation is entirely from https://github.com/pytorch/pytorch/pull/62964

albanD replied:
Why is this an nn.Module? This should be just a function no?

rohan-varma replied:
It is indeed an nn.Module but it is basically called as a function in L223. Do you mean this should be a torch.autograd.Function?

The implementation is directly from https://github.com/pytorch/pytorch/pull/62964/files, do we want to refactor this to be a function or stick with the original implementation?

albanD replied:
I meant it could be a plain old python function.
This is not going to be public API anyways.

zhaojuanmao commented on test/test_autograd.py:
```diff
@@ -26,7 +26,7 @@
 from torch.autograd.profiler_util import (_format_time, EventList, FunctionEvent, FunctionEventAvg)
 import torch.autograd.functional as autogradF
-from torch.utils.checkpoint import checkpoint
+from torch.utils.checkpoint import checkpoint, Checkpoint
```
nit: lint error

zhaojuanmao reviewed (commented):
nice!

rohan-varma committed:
Update on "[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd"


Resubmission of https://github.com/pytorch/pytorch/pull/62964 with the
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, the tests that we need to add are:

- [x] Gradient hooks are called once
- [ ] works when input does require grads but Tensor that require grads are captures (like first layer in a nn)
- [x] works for functions with arbitrary input/output objects
- [ ] distributed tests

Note that this is only for `torch.utils.checkpoint`, if this approach overall looks good, we will do something similar for `checkpoint_sequential`. 


Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

[ghstack-poisoned]

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -142,7 +142,7 @@ def backward(ctx, *args):
 
 
-def checkpoint(function, *args, **kwargs):
+def checkpoint(function, *args, use_reentrant: bool = True, **kwargs):
```
Maybe make it kwarg only? Or the fact that it is placed after `*args` already makes it kwarg only?

rohan-varma replied:
Yep, it is kwarg only since it is after *args: https://www.vegardstikbakke.com/python-keyword-only/

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -166,9 +166,11 @@ def checkpoint(function, *args, **kwargs):
-        Checkpointing currently only supports :func:`torch.autograd.backward`
-        and only if its `inputs` argument is not passed. :func:`torch.autograd.grad`
-        is not supported.
+        If ``use_reentrant=True`` is specified, checkpointing currently only
```
I feel like we could re-order some of the warnings. Especially because most of them only apply to the use_reentrant=True case.

rohan-varma replied:
Moved this warning to the bottom then, and still kept the wording as use_reentrant=True is the default for now. Let me know if they could be structured better.

albanD commented on test/test_autograd.py:
```diff
@@ -4308,6 +4308,35 @@ def test_checkpointing(self):
+        for r in range(num_inp):
+            data_r = torch.empty(1, nz_inp)
+            data_r.uniform_()
+            data_r.requires_grad = True
```
You can remove this to test the case where no input requires grad btw.

albanD commented on test/test_autograd.py:
```diff
@@ -4318,6 +4347,116 @@ def test_checkpoint_valid_reset_on_error(self):
+        Verifies gradient correctness when checkpoint without reentrant autograd
+        is used in conjunction with DataParallel.
+        """
+        class ExpModule(torch.nn.Module):
```
Should we use a Linear here maybe? So that the Module has some state?

albanD commented on test/test_autograd.py:
```diff
@@ -4318,6 +4347,116 @@ def test_checkpoint_valid_reset_on_error(self):
+            def forward(self, inp):
+                return torch.exp(inp)
+
+        a = torch.randn(20, 20, requires_grad=True)
```
nit: could be 2,2

albanD commented on test/test_autograd.py:
```diff
@@ -4318,6 +4347,116 @@ def test_checkpoint_valid_reset_on_error(self):
+
+        w.register_hook(hook)
+        x = torch.rand(10, 10, requires_grad=True)
+        h = w * x  # Using w outisde the checkpoint
```
outside

albanD reviewed (commented):
Sounds mostly good!
I think we can improve the doc a little to make sure people are more inclined to use this new version.

rohan-varma committed:
Update on "[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd"


Resubmission of https://github.com/pytorch/pytorch/pull/62964 with the
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, the tests that we need to add are:

- [x] Gradient hooks are called once
- [ ] works when input does require grads but Tensor that require grads are captures (like first layer in a nn)
- [x] works for functions with arbitrary input/output objects
- [ ] distributed tests

Note that this is only for `torch.utils.checkpoint`, if this approach overall looks good, we will do something similar for `checkpoint_sequential`. 


Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

[ghstack-poisoned]

rohan-varma commented:
@albanD Updated per your suggestions, although not completely sure about whether `class Checkpoint` should be a function or nn.Module. Could you take another look?

rohan-varma committed:
Update on "[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd"


Resubmission of https://github.com/pytorch/pytorch/pull/62964 with the
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, the tests that we need to add are:

- [x] Gradient hooks are called once
- [ ] works when input does require grads but Tensor that require grads are captures (like first layer in a nn)
- [x] works for functions with arbitrary input/output objects
- [ ] distributed tests

Note that this is only for `torch.utils.checkpoint`, if this approach overall looks good, we will do something similar for `checkpoint_sequential`. 


Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd"


Resubmission of https://github.com/pytorch/pytorch/pull/62964 with the
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, the tests that we need to add are:

- [x] Gradient hooks are called once
- [ ] works when input does require grads but Tensor that require grads are captures (like first layer in a nn)
- [x] works for functions with arbitrary input/output objects
- [ ] distributed tests

Note that this is only for `torch.utils.checkpoint`, if this approach overall looks good, we will do something similar for `checkpoint_sequential`. 


Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

[ghstack-poisoned]

rohan-varma commented on torch/utils/checkpoint.py:
```diff
@@ -190,6 +186,13 @@ def checkpoint(function, *args, **kwargs):
         model won't have gradients. At least one of the outputs needs to have
         :code:`requires_grad=True` as well.
 
```
Sounds good, I'll update this and add the appropriate tests as well

rohan-varma committed:
Update on "[Autograd/Checkpoint] Checkpoint implementation without reentrant autograd"


Resubmission of https://github.com/pytorch/pytorch/pull/62964 with the
suggestions and tests discussed in
https://github.com/pytorch/pytorch/issues/65537.

Adds a `use_reentrant=False` flag to `checkpoint` function. When
`use_reentrant=True` is specified, a checkpointing implementation that uses
SavedVariableHooks instead of re-entrant autograd is used. This makes it more
composable with things such as `autograd.grad` as well as DDP (still need to
add thorough distributed testing).

As discussed in https://github.com/pytorch/pytorch/issues/65537, the tests that we need to add are:

- [x] Gradient hooks are called once
- [x] works when input does require grads but Tensor that require grads are captures (like first layer in a nn)
- [x] works for functions with arbitrary input/output objects
- [x] distributed tests (next PR)

Note that this is only for `torch.utils.checkpoint`, if this approach overall looks good, we will do something similar for `checkpoint_sequential`. 


Differential Revision: [D32704467](https://our.internmc.facebook.com/intern/diff/D32704467/)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM
Thanks for sending the PR!

suo commented:
This unfortunately broke master (looks like a related test). https://hud.pytorch.org/commit/pytorch/pytorch/e032dae32904c6e90353e90051785514d6c5a7d9. I am reverting

rohan-varma commented:
@suo Thanks for the revert, the failure makes sense and unfortunately it didn't run on the PR because it was marked as slowTest .
